tools/shared.py

```diff
@@ -884,9 +884,12 @@ def get_llvm_target():
                                  '-D__EMSCRIPTEN_tiny__=' + str(EMSCRIPTEN_VERSION_TINY)]
 
 if LLVM_TARGET == WASM_TARGET:
-  # wasm target does not automatically define emscripten stuff, so do it here
+  # wasm target does not automatically define emscripten stuff, so do it here.
+  # Also, set the ABI version to at least 2 so that we get nicely aligned string
+  # data and other nice fixes.
   COMPILER_OPTS = COMPILER_OPTS + ['-DEMSCRIPTEN',
-                                   '-D__EMSCRIPTEN__']
+                                   '-D__EMSCRIPTEN__',
+                                   '-D_LIBCPP_ABI_VERSION=2']
```

Comment:
The only risk is if someone builds some code as separate modules with different versions of emscripten? We don't really make guarantees about that, our linking has changed quite a bit over time. So I think it's fine to do this change for everything.